scripts/internal_links_monitor.py

```diff
@@ -0,0 +1,243 @@
+                if port_info:
+                    return port_info['oper_status']
+        return 'down'
+
+    def filter_for_backend_ports(self, namespace, port_map):
+        ''' Returns only internal (backend) ports which are operationally up for monitoring '''
+        filtered_port_map = {}
+        for k,v in port_map.items():
+            if k.startswith('Ethernet-BP'):                
+                if self.get_port_status(namespace, k) == 'up':
+                    filtered_port_map[k] = v
+        return filtered_port_map
+
+    def get_lag_name_for_port(self, namespace, port_name):
+        ''' Returns name of the portchannel in which the given port is a member '''
+        lag_member = self.appdb[namespace].keys('APPL_DB', 'LAG_MEMBER_TABLE:*:{}'.format(port_name))
+        if len(lag_member):
+            table = lag_member[0].find(":") + 1
+            port = lag_member[0].find(":", table)
+            return lag_member[0][table:port]
+        return ''
+
+    def get_min_links_for_lag(self, namespace, lag_name):
+        ''' Returns min_links configuration of the given portchannel '''
+        lag_info = self.configdb[namespace].get_all('CONFIG_DB', 'PORTCHANNEL|{}'.format(lag_name))        
+        return int(lag_info['min_links'])        
+
+    def get_active_lag_member_count(self, namespace, lag_name):
+        ''' Returns number of member ports that are operationally up in the given portchannel '''
+        lag_members = self.appdb[namespace].keys('APPL_DB', 'LAG_MEMBER_TABLE:{}:*'.format(lag_name))
+        active_count = 0
+        for member in lag_members:
+            port = member.rsplit(':', 1)[1]            
+            if self.get_port_status(namespace, port) == 'up':
+                active_count += 1
+        return active_count
+
+    def isolate_lag_member(self, namespace, port_name):
+        ''' Set admin_status of given port to down in the CONFIG_DB '''
+        ret = self.configdb[namespace].set('CONFIG_DB', 'PORT|{}'.format(port_name), 'admin_status', 'down')
+        if ret == 0:
+            syslog.syslog(syslog.LOG_CRIT,
+                                        'Internal port {} has been shutdown to mitigate errors'.format(port_name))
+        else:
+            syslog.syslog(syslog.LOG_CRIT,
+                                        'Unable to shutdown internal port {}, return code {}'.format(port_name, ret))
+    def attempt_to_mitigate_ports(self, namespace, err_port_per_ns):
+        ''' Attempt to isolate given list of ports 
+            1. Check if port is already down, then skip mitigation
+            2. Check if number of active links in the portchannel where the port is a member is greater
+                than min_links. 
+                a. If active links is greater, shutdown the port
+                b. If not, return after generating syslog
+        '''
+
+        for port_name in err_port_per_ns:            
+            if self.get_port_status(namespace, port_name) == 'down':
```

Comment:
this can happen if the internal links are flapping, because we collect only the `up` ports in the port_map?, should we proceed with shutdown on the ports iso of ignoring the port?